src/librustc_mir/monomorphize/collector.rs

```diff
@@ -475,7 +484,9 @@ fn check_type_length_limit<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
     //
     // Bail out in these cases to avoid that bad user experience.
     let type_length_limit = *tcx.sess.type_length_limit.get();
-    if type_length > type_length_limit {
+    // We include the const length in the type length, as it's better
+    // to be overly conservative.
+    if type_length + const_length > type_length_limit {
```

Comment:
I think this is unnecessary, `.types()` / `.consts()` shouldn't be used on the `substs`, instead the substs should be uniformly "walked" somehow, only ignoring lifetimes.